stellargraph/connector/neo4j/graph.py

```diff
@@ -51,32 +71,139 @@ def nodes(self):
+                """
+            result = self.graph_db.run(
+                feature_query, parameters={"node_id_list": nodes}
+            )
+            rows = result.data()
+
+            # this method currently doesn't handle any other invalid IDs. If there are other
+            # invalid ids, we should raise an error
+            if len(rows) != sum(valid):
+                ids = {node["node_data"]["ID"] for node in rows}
+                invalid = [
+                    node for node in nodes if node is not None and node not in ids
+                ]
+                raise ValueError(
+                    f"nodes: Found values that did not return any results from the database: {comma_sep(invalid)}"
+                )
+
+            features[valid, :] = [row["node_data"]["features"] for row in rows]
+
+            return IndexedArray(features, index=nodes)
+
+    def node_features(self, nodes):
         """
         Get the numeric feature vectors for the specified nodes or node type.
 
         Args:
-            nodes (list): list of node IDs.
+            nodes (list or hashable, optional): Node ID or list of node IDs.
         Returns:
             Numpy array containing the node features for the requested nodes.
         """
-        feature_query = f"""
-            UNWIND $node_id_list AS node_id
-            MATCH(node) WHERE node.ID = node_id
-            RETURN node.features as features
-            """
-        result = self.graph_db.run(
-            feature_query, parameters={"node_id_list": node_ids},
-        )
-        features = np.array([row["features"] for row in result.data()])
-        return features
+        if self._nodes is not None:
+            return extract_element_features(
+                self._nodes,
+                self.unique_node_type,
+                "nodes",
+                nodes,
+                type=None,
+                use_ilocs=False,
+            )
+        else:
+            return self._node_features_from_db(nodes).values
+
+    def unique_node_type(self):
+        return self._node_type
+
+    def node_feature_sizes(self):
+        """
+        Get the feature sizes for the node types in the graph.
+
+        This method obtains the feature size by sampling a random node from the graph. Currently
+        this class only supports a single default node type, and makes the following assumptions:
+
+        - all nodes have features as a single list
+
+        - all nodes' features have the same size
+
+        Returns:
+            A dictionary of node type and integer feature size.
+        """
+        if self._node_feature_size is None:
+            # if feature size is unknown, take a random node's features
+            feature_query = f"""
+                {self._match_node()}
+                RETURN node.features as features LIMIT 1
+                """
+            result = self.graph_db.run(feature_query)
+            self._node_feature_size = len(result.data()[0]["features"])
```

Comment:
Potentially we could optimise this slightly in two ways:
- by doing the length-ing in the database and so send an integer, not the potentially-large feature vectors over the network
- avoiding create the `list` and `dict` structures with `run`+`data` when we're fetching a single result, since Py2neo has a special case function for that operation https://py2neo.org/v4/database.html#py2neo.database.Transaction.evaluate (https://py2neo.org/v4/database.html#py2neo.database.Cursor says "For queries that are expected to return only a single value within a single record, use the evaluate() method. This will return the first value from the next record or None if neither the field nor the record are present")

```suggestion
                RETURN size(node.features) LIMIT 1
                """
            self._node_feature_size = self.graph_db.evaluate(feature_query)
```